novawallet/Common/Services/AssetHubMigrationInfo/AHMInfoService.swift

```diff
@@ -0,0 +1,272 @@
+        self.applicationHandler = applicationHandler
+        self.ahmInfoRepository = ahmInfoRepository
+        self.assetBalanceRepository = assetBalanceRepository
+        self.settingsManager = settingsManager
+        self.operationQueue = operationQueue
+        self.workingQueue = workingQueue
+
+        super.init(logger: logger)
+    }
+}
+
+// MARK: - Private
+
+private extension AHMInfoService {
+    func createSetupWrapper() -> CompoundOperationWrapper<Void> {
+        let assetBalanceFetchOperation = assetBalanceRepository.fetchAllOperation(with: .init())
+
+        let resultOperation = ClosureOperation { [weak self] in
+            guard let self else { return }
+
+            let allBalances = try assetBalanceFetchOperation.extractNoCancellableResultData()
+
+            mutex.lock()
+
+            balances = allBalances.reduce(into: balances) { $0[$1.chainAssetId] = $1 }
+
+            mutex.unlock()
+        }
+
+        resultOperation.addDependency(assetBalanceFetchOperation)
+
+        return CompoundOperationWrapper(
+            targetOperation: resultOperation,
+            dependencies: [assetBalanceFetchOperation]
+        )
+    }
+
+    func createUpdatedConfigsWrapper() -> CompoundOperationWrapper<[AHMRemoteData]> {
+        let relevantConfigsWrapper = createRelevantConfigsWrapper()
+
+        let migrationTimestampsOperation = createMigrationTimestampsOperation(
+            dependingOn: relevantConfigsWrapper
+        )
+
+        migrationTimestampsOperation.addDependency(relevantConfigsWrapper.targetOperation)
+
+        let resultOperation: BaseOperation<[AHMRemoteData]> = ClosureOperation {
+            let configs = try relevantConfigsWrapper.targetOperation.extractNoCancellableResultData()
+
+            let timestamps = try migrationTimestampsOperation
+                .extractNoCancellableResultData()
+                .compactMap { $0 }
+                .reduce(into: [:]) { $0[$1.0] = $1.1 }
+
+            return configs.filter {
+                guard let timestamp = timestamps[$0.sourceData.chainId] else { return false }
+
+                return Date(timeIntervalSince1970: TimeInterval(timestamp)) <= Date()
+            }
+        }
+
+        resultOperation.addDependency(migrationTimestampsOperation)
+
+        return CompoundOperationWrapper(
+            targetOperation: resultOperation,
+            dependencies: relevantConfigsWrapper.allOperations + [migrationTimestampsOperation]
+        )
+    }
+
+    func createRelevantConfigsWrapper() -> CompoundOperationWrapper<[AHMRemoteData]> {
+        let ahmInfoWrapper = ahmInfoRepository.fetchAllWrapper()
+
+        let resultOperation = ClosureOperation<[AHMRemoteData]> { [weak self] in
+            guard let self else { return [] }
+
+            let shownChains = settingsManager.ahmInfoShownChains
+
+            let ahmInfoConfigs = try ahmInfoWrapper.targetOperation.extractNoCancellableResultData()
+
+            mutex.lock()
```

Comment:
WDYT if we have setBalances and fetchBalanceById functions and use locks only inside them?